scripts for immune deconvolution

komalsrathi opened the pull request:
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->

### Purpose/implementation Section

#### What scientific question is your analysis addressing?

Immune profiling of PBTA histologies.

#### What was your approach?

To deconvolute the immune cell types across 20 histologies from the PBTA cohort, we use the R package immunedeconv, which is a unified interface that allows cell type quantification from RNA-seq gene expression data using multiple methods i.e. EPIC (n = 6), TIMER (n = 6), MCP-counter (n = 8), quanTIseq (n = 10), CIBERSORT (n = 22) and xCell (n = 64). Out of the six methods, we chose the top two most comprehensive methods i.e. CIBERSORT and xCell that are able to deconvolute 22 and 64 cell types, respectively. In a benchmarking study, xCell was also shown to robustly outperform all other methods, including CIBERSORT. 

In general, immune deconvolution is a hard problem in that most immune deconvolution methods do not always agree because of the type of underlying algorithm and ability to deconvolute certain cell types better than others. To compare the outcome scores across these two prominent profiling methods, we took 13 cell types common between both methods and created a correlation plot per cell type per histology. 

In order to see histology specific enrichment of certain cell types, we created heatmaps for xCell and CIBERSORT representing average immune cell scores normalized across histologies. 

#### What GitHub issue does your pull request address?

https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/15

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?

#### Is there anything that you want to discuss further?

Any suggestions are most welcomed.

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes

### Results

#### What types of results are included (e.g., table, figure)?

01-immune-deconv.R creates an RData containing immune scores corresponding to xCell and CIBERSORT. 

02-summary-plots.R creates a correlation plot between cell types common between xCell and CIBERSORT as well as heatmaps containing average immune scores per cell type per histology for each method.

#### What is your summary of the results?

Correlation plot:
Overall, across all histologies and cell types, the predictions did not correlate well (pearson correlation 0.12) but we did observe a high correlation (> 0.5) between certain cell type predictions i.e. Macrophages M2, Monocytes, Neutrophils and T cell CD8+ markers across specific histologies. 

Heatmaps:
We observed a very similar and interesting pattern in the heatmaps as well i.e. cell types that were highly correlated between xCell and CIBERSORT are also the ones that found in high proportions compared to other cell types. Using the profiling method, we could distinctively identify three histologies enriched in specific cell types. Histiocytic tumors (n = 5) which are known to be characterized by an increase in Monocytes, Macrophages and Dendritic cells[3] were observed to be enriched in Myeloid dendritic cell activated, Monocytes, Neutrophils and Macrophage M1/M2; Lymphomas (n = 1) were seen to be enriched in B cell plasma, Class-switched memory B cell, B cell naive, B cell, B cell memory, T cell regulatory (Tregs), Common Lymphoid progenitor[4]; and Germ cell tumors (n = 13) as expected[5] were seen to be enriched in T cell CD8+ and T cell CD4+ naive cells.

Because we only have a sample size of 1 for Lymphomas, we couldn't get any correlation between immune scores of xCell and CIBERSORT across various B cell types but looking at the heatmap of average scores per cell type per histology, it is highly likely that those two would correlate. 

#### Reproducibility Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

jaclyn-taroni commented:
Hi @komalsrathi, thank you for filing this! We are probably unlikely to provide a full review prior to the holiday this week, thanks for your patience.

I noticed that you are adding the CIBERSORT R script here (`analyses/immune-deconv/CIBERSORT.R`). Unfortunately, I believe that the CIBERSORT licensing terms are incompatible with the license we use for this repository, which is BSD-3 for source code. Tagging @cgreene who has run into this issue with CIBERSORT before if I recall correctly.

If this is in fact the case, I'd recommend the following:

* Removing the `analyses/immune-deconv/CIBERSORT.R`/stop tracking in git and add this file to a `.gitignore` file in the `analyses/immune-deconv` directory.
* Making the CIBERSORT step optional such that you are able to skip it in continuous integration. This is required because when the repository gets checked out first thing in continuous integration `CIBERSORT.R` will not be available. This is probably relatively straightforward in `analyses/immune-deconv/01-immune-deconv.R`. (You may find the [**Passing variables only in CI** section of the README](https://github.com/AlexsLemonade/OpenPBTA-analysis#passing-variables-only-in-ci) helpful depending on how you choose to implement this.) I have two thoughts about how to possibly deal with this in `analyses/immune-deconv/02-summary-plots.R`: 
  * Checking for the existence of the CIBERSORT results before making the correlation plot and the CIBERSORT heatmap. Because we also want to ensure that the correlation plot steps execute in continuous integration, you could check for correlation between xCell and another method specified as an option.
  * Because you're adding `analyses/immune-deconv/deconv-output.RData` as part of this pull request, you could have logic in `analyses/immune-deconv/01-immune-deconv.R` where you check for the existence of the `analyses/immune-deconv/deconv-output.RData` file and _only_ overwrite it if specified as an option. You would elect not to overwrite the file in CI and then it would use this one that you've committed in `02-summary-plots.R` as is.

Looking forward to your thoughts on implementation if we need to work around CIBERSORT.

Thank you again for the contribution and have a happy holiday break!

komalsrathi commented:
> Hi @komalsrathi, thank you for filing this! We are probably unlikely to provide a full review prior to the holiday this week, thanks for your patience.
> 
> I noticed that you are adding the CIBERSORT R script here (`analyses/immune-deconv/CIBERSORT.R`). Unfortunately, I believe that the CIBERSORT licensing terms are incompatible with the license we use for this repository, which is BSD-3 for source code. Tagging @cgreene who has run into this issue with CIBERSORT before if I recall correctly.
> 
> If this is in fact the case, I'd recommend the following:
> 
> * Removing the `analyses/immune-deconv/CIBERSORT.R`/stop tracking in git and add this file to a `.gitignore` file in the `analyses/immune-deconv` directory.
> * Making the CIBERSORT step optional such that you are able to skip it in continuous integration. This is required because when the repository gets checked out first thing in continuous integration `CIBERSORT.R` will not be available. This is probably relatively straightforward in `analyses/immune-deconv/01-immune-deconv.R`. (You may find the [**Passing variables only in CI** section of the README](https://github.com/AlexsLemonade/OpenPBTA-analysis#passing-variables-only-in-ci) helpful depending on how you choose to implement this.) I have two thoughts about how to possibly deal with this in `analyses/immune-deconv/02-summary-plots.R`:
>   
>   * Checking for the existence of the CIBERSORT results before making the correlation plot and the CIBERSORT heatmap. Because we also want to ensure that the correlation plot steps execute in continuous integration, you could check for correlation between xCell and another method specified as an option.
>   * Because you're adding `analyses/immune-deconv/deconv-output.RData` as part of this pull request, you could have logic in `analyses/immune-deconv/01-immune-deconv.R` where you check for the existence of the `analyses/immune-deconv/deconv-output.RData` file and _only_ overwrite it if specified as an option. You would elect not to overwrite the file in CI and then it would use this one that you've committed in `02-summary-plots.R` as is.
> 
> Looking forward to your thoughts on implementation if we need to work around CIBERSORT.
> 
> Thank you again for the contribution and have a happy holiday break!

@jaclyn-taroni Thank you for your feedback. I agree, we shouldn't have the CIBERSORT.R and LM22.txt available (my bad for overlooking that). I am going to think about this a little more and get back with my thoughts after the break. Have a Happy Thanksgiving!

cgreene commented:
CIBERSORT's license used to be... extensive. @dhimmel made a copy of it here:
https://gist.github.com/dhimmel/58dcd9b512e669f20a65ddf73997b733

Is the license the same or have they modified it? It seems plausible that they could have updated it so that you could add a command to download it to the CI system with wget if there's a URL where it's available.

jaclyn-taroni commented:
Ah, okay it does look like it's been updated @cgreene, I am copying the text from the modal that comes up when you click ` Non-Commercial Terms of Use` on the [About page](https://cibersort.stanford.edu/about.php) (I do not have an account):


> The Board of Trustees of the Leland Stanford Junior University (“Stanford”) provides CIBERSORT website features and services (“Service”) free of charge for non-commercial use only. Use of the Service by any commercial entity for any purpose, including research, is prohibited.
> 
> By using the Service, you agree to be bound by the terms of this Agreement. Please read it carefully.
> 
>     You agree not to use the Service for commercial advantage, or in the course of for-profit activities. You agree not to use the Service on behalf of any organization that is not a non-profit organization. Commercial entities wishing to use this Service should contact Stanford University’s Office of Technology Licensing and reference docket S13-279.
>     You agree not to use the Service for diagnosis, treatment, cure, prevention, or mitigation of disease or other conditions in man or other animals. You acknowledge that the Service and any information obtained therefrom is not intended to substitute for care by a licensed healthcare professional.
> 
> THE SERVICE IS OFFERED “AS IS”, AND, TO THE EXTENT PERMITTED BY LAW, STANFORD MAKES NO REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED. STANFORD SHALL NOT BE LIABLE FOR ANY CLAIMS OR DAMAGES WITH RESPECT TO ANY LOSS OR OTHER CLAIM BY YOU OR ANY THIRD PARTY ON ACCOUNT OF, OR ARISING FROM THE USE OF THE SERVICE. YOU HEREBY AGREE TO DEFEND AND INDEMNIFY STANFORD, ITS TRUSTEES, EMPLOYEES, OFFICERS, STUDENTS, AGENTS, FACULTY, REPRESENTATIVES, AND VOLUNTEERS (“STANFORD INDEMNITEES”) FROM ANY LOSS OR CLAIM ASSERTED AGAINST STANFORD INDEMNITEES ARISING FROM YOUR USE OF THE SERVICE.
> 
>     All rights not expressly granted to you in this Agreement are reserved and retained by Stanford or its licensors or content providers. This Agreement provides no license under any patent.
>     You agree that this Agreement and any dispute arising under it is governed by the laws of the State of California, United States of America, applicable to agreements negotiated, executed, and performed within California.
> 
> Subject to your compliance with the terms and conditions set forth in this Agreement, Stanford grants you a revocable, non-exclusive, non-transferable right to access and make use of the Service.

jaclyn-taroni commented:
Hi @komalsrathi, I hope you had a great break! I wanted to check in. I have not requested or provided a review of this yet because of the expected CIBERSORT changes. Do you have an idea of when you will have the opportunity to revisit this pull request? Thank you!

komalsrathi commented:
> Hi @komalsrathi, I hope you had a great break! I wanted to check in. I have not requested or provided a review of this yet because of the expected CIBERSORT changes. Do you have an idea of when you will have the opportunity to revisit this pull request? Thank you!

@jaclyn-taroni I should have the necessary changes by end of next week.

jaclyn-taroni commented:
Sounds good, thanks!

komalsrathi commented:
@jaclyn-taroni I am working on this today, will let you know when this is ready for review.

komalsrathi commented:
@jaclyn-taroni I have made the changes that you suggested:

1. Added CIBERSORT.R and LM22.txt to `analyses/immune-deconv/.gitignore`
2. Use `xCell` as method 1 because it has the most immune cell types to deconvolute and performs better than others (I have added citations in corresponding methods) and pass the second method as a variable. 
3. For the second method, the script will look for the env variable `OPENPBTA_DECONV_METHOD`. If it is set to `cibersort_abs`, then the script will look for CIBERSORT.R and LM22.txt files under `analyses/immune-deconv` and run CIBERSORT. 
4. If `OPENPBTA_DECONV_METHOD` is not found, then the second method is set to `mcp_counter` (for MCP-counter). This is one of the three best methods (along with xCell, CIBERSORT) found in a benchmarking [study](https://cancerres.aacrjournals.org/content/79/24/6238)

Please review when you can and let me know any changes/suggestions. Thank you!!

jaclyn-taroni commented:
The `CIBERSORT.R` and `LM22.txt` files are ignored but they have not been removed/deleted yet. I will delete these at the same time as updating this branch to be in sync with `master` and then provide a review.

komalsrathi commented on analyses/immune-deconv/01-immune-deconv.R:
```diff
@@ -0,0 +1,96 @@
+common.genes <- intersect(rownames(polya), rownames(stranded))
+polya <- polya[common.genes,]
+stranded <- stranded[common.genes,]
+expr.input <- cbind(polya, stranded)
```
@jaclyn-taroni Quoting a few sentences from two papers that I referred to before merging all the samples together:

```
Comprehensive evaluation of transcriptome-based cell-type quantification methods for immuno-oncology: PMID 31510660
1. xCell uses the variability among the samples for a linear transformation of the output score.
2. xCell does not detect any signal when ran with few non-heterogeneous samples
```

```
xCell: digitally portraying the tissue cellular heterogeneity landscape: PMID 29141660

1. We chose to apply a gene signature enrichment approach over deconvolution methods because of several advantages that the former provides. 
First, gene signatures are rank-based and are therefore suitable for cross-platform transcriptome measurements. 
We showed here that our scores reliably predict enrichment when using different RNA-seq techniques and different microarray platforms. 
They are agnostic to normalization methods or concerns related to batch effects, making them robust to both technical and biological noise.
```

cgreene commented:
@komalsrathi : can you help us understand why you needed to force push here:
> komalsrathi force-pushed the komalsrathi:immune-deconv branch from c79de3a to 401b9ad 37 minutes ago

Do you know what changes were wiped out by this? I'm guessing they were edits that were made by others or on GitHub.

komalsrathi commented:
I was trying to pull the changes that I thought had been made to this
branch - can this be fixed?

On Fri, Jan 3, 2020 at 4:59 PM Casey Greene <notifications@github.com>
wrote:

> @komalsrathi <https://github.com/komalsrathi> : can you help us
> understand why you needed to force push here:
>
> komalsrathi force-pushed the komalsrathi:immune-deconv branch from c79de3a
> <https://github.com/AlexsLemonade/OpenPBTA-analysis/commit/c79de3afe4697a77d99aba083bd7ee8b62af4ac7>
> to 401b9ad
> <https://github.com/AlexsLemonade/OpenPBTA-analysis/commit/401b9ad66bfc4471c8f541c7b4467ff9697ca4f7>
> 37 minutes ago
>
> Do you know what changes were wiped out by this? I'm guessing they were
> edits that were made by others or on GitHub.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/292?email_source=notifications&email_token=ABVNEJ7A7SUWXKWEGKLPOWLQ36YNPA5CNFSM4JR4IALKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEICFVXA#issuecomment-570710748>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVNEJ3K4XBXS3ADNTEPPTDQ36YNPANCNFSM4JR4IALA>
> .
>
-- 
*Komal S Rathi* | Bioinformatics Scientist II, DBHi, The Children's
Hospital of Philadelphia | rathik@email.chop.edu

cgreene commented:
@komalsrathi Just corresponded a bit with @jaclyn-taroni. Fortunately she checked it out locally so she could force push back to the state things were at when she was last making changes.

How many changes were you intending to make (vs. how many of the changes from the commits came in from the merge).

komalsrathi commented:
@cgreene I guess I was just trying to pull changes that Jaclyn had made here to my local branch: in order to get all changes on my local system and then work on her other suggestions. Because there were some changes that I had accepted for e.g. where she had asked to create a plots directory.

cgreene commented:
Ok! I see merge markers so it sounds like you pulled, there were merge conflicts, and then when you tried to push it didn't work b/c it wasn't yet merged. I will let @jaclyn-taroni propose next steps on this, but I think her force pushing and you re-integrating those changes will be easier than any other alternative. Thanks for your help understanding what the goal was here! It'll help us get the best path to a successful merge. 😁

komalsrathi commented:
> Ok! I see merge markers so it sounds like you pulled, there were merge conflicts, and then when you tried to push it didn't work b/c it wasn't yet merged. I will let @jaclyn-taroni propose next steps on this, but I think her force pushing and you re-integrating those changes will be easier than any other alternative. Thanks for your help understanding what the goal was here! It'll help us get the best path to a successful merge. 😁

I apologize for this, I don't have a full understanding of some of the commands and I knew I had done something weird when I force pushed. I will wait for instructions before trying anything else. Thank you!

jaclyn-taroni commented:
Yes, I am planning to force push the changes up to the point before the suggested changes were applied via the GitHub interface. Then you're able to commit the suggested changes that are marked as resolved/that you intended include @komalsrathi 

Once that's done, I would pull the changes locally with:

```
# checkout the branch for this pull request
git checkout immune-deconv
# pull the changes locally
git pull
```

If that's successful, you can make, add, and commit changes locally. If you try those steps are they are not successful or you get an error when you try to push, can you post here or in Slack with what you're seeing locally and we can get to the bottom of it.

jaclyn-taroni commented:
Okay - that looks like it worked and removed all the CIBERSORT files. But now this branch is out of date with the `master` branch and we have conflicts in the Dockerfile that need to be resolved. This is probably where there was trouble pulling locally. I am going to resolve the conflicts via the GitHub interface which will create a new merge commit.

jaclyn-taroni commented:
The suggestions that were applied earlier are marked as resolved so they can no longer be applied. I will mark them as unresolved so we're able to apply the changes via the GitHub interface. That will get us back to the point where I submitted my review.

jaclyn-taroni commented:
Alright! You are all set to apply my suggestions via the GitHub interface. Once you've applied the suggestions, you essentially want to get the same state of the branch on your local machine as what's here remotely. You could do what is called a `reset` in git parlance to overwrite what you have locally (should take care of the merge conflicts). 

This [StackOverflow post](https://stackoverflow.com/questions/1125968/how-do-i-force-git-pull-to-overwrite-local-files) has a lot of good information in it and is what I referenced while writing this up.

The command for that is (**warning this will get rid of _any_ local changes!**):

```
git reset --hard origin/immune-deconv
```

I am happy to walk you through all of that when we meet in person, but I wanted to document this while it was fresh in my mind.

komalsrathi commented:
@jaclyn-taroni Hi Jaclyn, I think this should be ok now. This is what I did:

```
git fetch origin immune-deconv
git reset --hard origin/immune-deconv
```

Then, I applied the changes via Github interface and again:

```
git fetch origin immune-deconv
git reset --hard origin/immune-deconv
```

This is the log on my local system:
```
➜  OpenPBTA-analysis git:(immune-deconv) git log --oneline
b75cd7f (HEAD -> immune-deconv, origin/immune-deconv) Update analyses/immune-deconv/run-immune-deconv.sh
```

Now, I just need to make and commit other changes that you have suggested in the review. Can you comment, if this sounds right?

jaclyn-taroni commented:
> Now, I just need to make and commit other changes that you have suggested in the review. Can you comment, if this sounds right?

Yep, this looks correct 👍 thanks!

komalsrathi commented:
> Hi @komalsrathi,
> 
> Thanks for adding this analysis! I've added specific comments aimed at bringing the folder organization more in line with the rest of the repository and using `rprojroot` to head off any potential issues with sourcing your plot theme file. In addition to the line comments I've added, I have some more general comments and questions about this analysis:
> 
> * Can you add this module to the ["modules at a glance" table in `analyses/README.md`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses#modules-at-a-glance) please? That was added as a requirement since this pull request was initiall filed ([ref](https://github.com/AlexsLemonade/OpenPBTA-analysis#documenting-your-analysis)).

Done

> * Similarly, can you add a README to `immune-deconv` here? I think this addition will help with some of the outstanding questions I have (see below).

Here: https://github.com/komalsrathi/OpenPBTA-analysis/blob/immune-deconv/analyses/immune-deconv/README.md

> * How do you intend for this module to be used? Specifically, do you expect that folks will run only two methods (e.g., `xCell` and `cibersort_abs`) or is your intention to make this flexible enough that people can run multiple methods that they can compare downstream? I'm asking in part because you made reference to creating correlation plots for all methods in the past [#15 (comment)](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/15#issuecomment-548442448).

Please see the output section of: https://github.com/komalsrathi/OpenPBTA-analysis/blob/immune-deconv/analyses/immune-deconv/README.md#choice-of-method

> * How do you intend for folks to use the results of this analysis, e.g., `results/deconv-output.RData`, downstream? It looks like there is [a section of the `immunedeconv` documentation about the interpretation of scores](https://icbi-lab.github.io/immunedeconv/articles/immunedeconv.html#interpretation-of-scores) that should probably be mentioned in the README. It looks like you have some information around this in your methods pull request [AlexsLemonade/OpenPBTA-manuscript#64](https://github.com/AlexsLemonade/OpenPBTA-manuscript/pull/64)

https://github.com/komalsrathi/OpenPBTA-analysis/blob/immune-deconv/analyses/immune-deconv/README.md#01-immune-deconvr

> * You're combining the two RNA-seq matrices that have been split up because we observed technical effects. How does combining the matrices (or not) affect your results?

MCP_Counter gives exactly the same result for polyA and stranded, if the matrix is combined or not. CIBERSORT (abs.) also gives the same result for polyA - whether combined or not. But for stranded data, there is some loss of information when using the combined matrix because polyA has lesser genes compared to stranded. In this case, if the stranded data is run independently, a few more cell types pop up. Similarly, for xCell not much changes when using the combined matrix vs the independent matrices (pearson correlation 0.99). So, to address the issue with CIBERSORT (abs.), I have changed `01-immune-deconv.R` such that we run the methods on both datasets independently and then combine the output. The results haven't changed much after applying this change.

> 
> Thanks again!

jaclyn-taroni commented:
Modules at a glance added with https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/292/commits/9fc1439c91ca4c6c309529b343d0aa39fea1571f

jaclyn-taroni commented on analyses/immune-deconv/README.md:
```diff
@@ -0,0 +1,64 @@
+## Immune Deconvolution
```
Discussed in person: Use one sentence per line same as manubot

komalsrathi replied:
Changed to one sentence per line: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/292/commits/10485cfeb6de79267be7331d26d93b96cc8e369c

komalsrathi commented on analyses/immune-deconv/01-immune-deconv.R:
```diff
@@ -0,0 +1,96 @@
+  return(res)
+}
+  
+# Deconvolute using xCell and Cibersort (absolute mode)
```
removed explicit mention of cibersort: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/292/commits/edc181d6e6fbdd81605a77a6bc5b592053527873

komalsrathi commented on analyses/immune-deconv/pubTheme.R:
```diff
@@ -0,0 +1,70 @@
+# Author: Komal S. Rathi
```
moved to util folder: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/292/commits/ab259b32652dad80cc5db844f228ee93b48055d2

komalsrathi commented on analyses/immune-deconv/02-summary-plots.R:
```diff
@@ -0,0 +1,123 @@
+suppressPackageStartupMessages(library(corrplot))
+
+# source plotting theme
+source('pubTheme.R')
```
using `rprojroot` here: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/292/commits/efa735d95c40ddd9e086dc294b7581a27b1a29c5

komalsrathi commented on .circleci/config.yml:
```diff
@@ -30,6 +30,16 @@ jobs:
+          name: Immune deconvolution and Summary plots using immunedeconv R package
+          command: ./scripts/run_in_ci.sh bash analyses/immune-deconv/run-immune-deconv.sh
+
+      # Immune deconvolution using CIBERSORT and xCell (for OpenPBTA)
```
removed commented section: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/292/commits/b2b1747ac993e312829c5f3416b27bc05b584bbd

komalsrathi commented:
@jaclyn-taroni, I have made the discussed changes. Let me know whenever you get a chance to review this. And thanks for helping out!
